src/modules/sensors/analog_battery.cpp

```diff
@@ -0,0 +1,140 @@
+#include "analog_battery.h"
+
+// Defaults to use if the parameters are not set
+#if BOARD_NUMBER_BRICKS > 0
+#if defined(BOARD_BATT_V_LIST) && defined(BOARD_BATT_I_LIST)
+static constexpr int   DEFAULT_V_CHANNEL[BOARD_NUMBER_BRICKS] = BOARD_BATT_V_LIST;
+static constexpr int   DEFAULT_I_CHANNEL[BOARD_NUMBER_BRICKS] = BOARD_BATT_I_LIST;
+#else
+static constexpr int   DEFAULT_V_CHANNEL[BOARD_NUMBER_BRICKS] = {0};
+static constexpr int   DEFAULT_I_CHANNEL[BOARD_NUMBER_BRICKS] = {0};
+#endif
+#else
+static constexpr int DEFAULT_V_CHANNEL[1] = {0};
+static constexpr int DEFAULT_I_CHANNEL[1] = {0};
+#endif
+
+AnalogBattery::AnalogBattery() :
+	ModuleParams(nullptr)
+{}
+
+void
+AnalogBattery::updateBatteryStatusRawADC(hrt_abstime timestamp, int32_t voltage_raw, int32_t current_raw,
+		bool selected_source, int priority, float throttle_normalized,
+		bool armed)
+{
+	// TODO: Check that there was actually a parameter update
+	_get_battery_base().updateParams();
+	updateParams();
+
+	float voltage_v = (voltage_raw * _get_cnt_v_volt()) * _get_v_div();
+	float current_a = ((current_raw * _get_cnt_v_curr()) - _get_v_offs_cur()) * _get_a_per_v();
+
+	bool connected = voltage_v > BOARD_ADC_OPEN_CIRCUIT_V &&
+			 (BOARD_ADC_OPEN_CIRCUIT_V <= BOARD_VALID_UV || is_valid());
+
+	_get_battery_base().updateBatteryStatus(timestamp, voltage_v, current_a, connected,
+						selected_source, priority, throttle_normalized, armed, false);
+
+	// Before refactoring and adding the BatteryBase and AnalogBattery classes, the only place that checked the
+	// value of the BAT_SOURCE parameter was in the ADC polling in sensors.cpp. So I keep that logic here for now.
+	if (_get_source() == 0) {
+		_get_battery_base().publish();
+	}
+}
+
+/**
+ * Whether the ADC channel for the voltage of this battery is valid.
+ * Corresponds to BOARD_BRICK_VALID_LIST
+ */
+bool AnalogBattery::is_valid()
+{
+#ifdef BOARD_BRICK_VALID_LIST
+	bool valid[BOARD_NUMBER_BRICKS] = BOARD_BRICK_VALID_LIST;
+	return valid[_get_brick_index()];
+#else
+	// TODO: Maybe return false instead?
```

Comment:
What sort of setup would that be? :thinking:

Reply:
I'm not sure. This is replicating the original mess of #if's in Sensors.cpp. I found that if I tried to simplify any of it, some builds would fail, so presumably the original author of that part of Sensors.cpp (I think @bkueng) had a good reason for this.

If you'd like, I can investigate this particular case, and find where it is necessary.

Reply:
I see. Leave as is then.